### PR TITLE
Encapsulate memos

### PR DESCRIPTION
**Pull request by sellout**
There’s a `Memo` class that was introduced with the WalletTxBuilder work, but it
was restricted to the wallet. This moves it to libzcash, uses it universally,
and makes it a lot more robust.

The only place (outside of tests) that uses `Memo::Bytes` is
`BaseNotePlaintext`, which needs a particular in-memory representation. But the
API for `*NotePlaintext` is in terms of the `Memo` class.

### CODE REVIEW

**Review by nuttycom — CHANGES_REQUESTED** [review]
A couple of blocking questions as to where we should be using the `Memo::Bytes` representation instead of the `Memo` type.

File: src/zcash/NoteEncryption.cpp [hunk 1]
```diff
@@ -133,7 +133,7 @@ std::optional<SaplingEncCiphertext> SaplingNoteEncryption::encrypt_to_recipient(
 
     crypto_aead_chacha20poly1305_ietf_encrypt(
         ciphertext.begin(), NULL,
-        message.begin(), ZC_SAPLING_ENCPLAINTEXT_SIZE,
+        message.begin(), SAPLING_ENCPLAINTEXT_SIZE,
```
[the remark this answers is not in the document]
Comment (sellout) [reply src/zcash/NoteEncryption.cpp:136]: So, some of the constants moved to NoteEncryption.hpp because of the new reliance on `Memo::SIZE`. In the move, it seemed like a good idea to change them from `#define` to `constexpr`, which can (and should) be namespaced. Namespacing them makes the `ZC_` redundant in favor of `libzcash::`.

I think that set of changes makes sense to do together. But I could make a commit that does the moving and renaming on its own, then apply the memo changes on top.

File: src/wallet/rpcwallet.cpp [hunk 2]
```diff
@@ -3271,11 +3269,11 @@ UniValue z_getaddressforaccount(const UniValue& params, bool fHelp)
             throw JSONRPCError(RPC_INVALID_PARAMETER, "diversifier index must be a decimal integer.");
         }
         auto parsed_diversifier_index = parsed_diversifier_index_opt.value();
-        if (parsed_diversifier_index.size() > ZC_DIVERSIFIER_SIZE) {
+        if (parsed_diversifier_index.size() > DIVERSIFIER_SIZE) {
```
[the remark this answers is not in the document]
Comment (sellout) [reply src/wallet/rpcwallet.cpp:3272]: See https://github.com/zcash/zcash/pull/6612#discussion_r1178126427

File: src/wallet/wallet.h [hunk 3]
```diff
@@ -381,7 +381,7 @@ struct SaplingNoteEntry
     SaplingOutPoint op;
     libzcash::SaplingPaymentAddress address;
     libzcash::SaplingNote note;
-    std::array<unsigned char, ZC_MEMO_SIZE> memo;
+    std::optional<libzcash::Memo> memo;
```
[the remark this answers is not in the document]
Comment (sellout) [reply src/wallet/wallet.h:384]: So, the updated `Memo` class can no longer hold a `0xF600…00` value. `std::optional<Memo>` means “a possibly-empty memo”.

Prior to this PR, both `std::nullopt` and `std::optional(Memo({{0xf6}}))` were representations of the empty memo, but they weren’t always checked consistently. And there were multiple constants defined as `{{0xf6}}`, as well as other places that `{{0xf6}}` was used directly as a magic array. So, a significant improvment in this PR is that `std::nullopt` is the only empty memo representation.

tl;dr – if we just had `libzcash::Memo memo;` here, then that wouldn’t allow us to represent the empty memo.

[the remark this answers is not in the document]
Comment (nuttycom) [reply src/wallet/wallet.h:384]: No; I'm suggesting that this be `Memo::Bytes`.

[the remark this answers is not in the document]
Comment (sellout) [reply src/wallet/wallet.h:384]: Ah, I see … IMO, we should stick with the “higher-level” wrapper as much as possible. Changing `SaplingNoteEntry` to  use `Bytes` leaks to the RPC layer. `BaseNotePlaintext` is the only thing that deals with serialization/encryption, so that’s where we use `Memo::Bytes`.

**Comment by sellout** [comment]
Commit-by-commit review is probably best.

**Review by nuttycom — APPROVED** [review]
utACK 601fa4b7c7583cbbbd325a9ae5c6886edc930869
